### ISSUE TITLE
False positive: Non-primitive type cast error if the function has a constrained lifetime

### ISSUE THREAD

**Issue by Jason5Lee** [rust-lang/rust-analyzer#20466]
**rust-analyzer version**: 0.3.2577-standalone (9db05508ed 2025-08-11)

**rustc version**: 1.88.0 (6b00bc388 2025-06-23)

**editor or extension**: Zed and VSCode (rust-analyzer extension)

**code snippet to reproduce**:

```rust
fn test2<'a>()
// If the following line is commented out, the false-positive error disappears, even though the lifetime has no connection to the error expression.
where 'a: 'static,
{
    let _: &[i32] = &[1i32] as &[_]; // False positive: non-primitive cast: &[i32; 1] as &[i32]
}
```

~~I’m uncertain whether this is the same as one of the "non-primitive cast" issues, as I can only reproduce it using macros—specifically, both `async-trait` and `log`. I’m unsure of the exact expanded code triggers this issue. Currently, I’ve found that if I attempt to log with a key-value pair inside an async function implemented for an async trait (using the `async-trait` crate rather than the built-in async trait mechanism), it produces a false positive error, whereas `cargo check` shows no issues.~~
I've identified the minimal reproducible example. The issue arises when a function has a constrained lifetime (whether bound to another lifetime or `'static`). In such cases, casting a borrow of an array to a borrow of a slice triggers a non-primitive cast error—even the lifetime is not related. However, if the lifetime is unconstrained (or absent entirely), the cast works without issues.

**Comment by evanj** [comment]
I ran into this using the `log` crate's `kv` feature in async functions. Example (not an async example: I think reproducing this with `async` needs some `awaits` and captured variables that I couldn't reproduce easily)

```rust
fn foo<'a>()
where
    'a: 'static,
{
    log::info!(key = 5; "log with kv"); // non-primitive cast: `&[(&'static str, Value<'_>); 1]` as `&[(&str, Value<'_>)]`
}
```

**Comment by techno-tanoC** [comment]
I also ran into this with the log crate and the async-trait crate.

```rust
#[async_trait::async_trait]
trait TestTrait {
    async fn test(&self);
}

struct TestStruct;

#[async_trait::async_trait]
impl TestTrait for TestStruct {
    async fn test(&self) {
        log::info!(status = "ok"; "hello"); // non-primitive cast: `&[(&'static str, Value<'_>); 1]` as `&[(&str, Value<'_>)]`rust-analyzer(E0605)
    }
}

#[tokio::main]
async fn main() {
    env_logger::init();

    TestStruct.test().await;
}

```

async-trait = "0.1.89"
log = { version = "0.4.28", features = ["kv"] }
VS Code rust-analyzer extension: v0.3.2593

**Comment by techno-tanoC** [comment]
In my case (the log crate with the kv feature and the async-trait crate), this issue did not reproduce with rust-analyzer extension v0.4.2608.

**Comment by ChayimFriedman2** [comment]
Yeah, the new trait solver probably fixed those issues, and the remaining will be fixed by https://github.com/rust-lang/rust-analyzer/pull/20664.

**Comment by ChayimFriedman2** [comment]
The issue is fixed today.
